geopandas/tools/sjoin.py

```diff
@@ -350,53 +417,25 @@ def _frame_join(join_df, left_df, right_df, how, lsuffix, rsuffix):
     right_index = right_df.columns[:right_nlevels]
 
     # perform join on the dataframes
-    if how == "inner":
-        join_df = join_df.set_index("_key_left")
-        joined = (
-            left_df.merge(join_df, left_index=True, right_index=True)
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                left_on="_key_right",
-                right_index=True,
-                suffixes=(None, None),
-            )
-            .drop(["_key_right"], axis=1)
-        )
-        joined = _restore_index(joined, left_index, left_index_original)
-
-    elif how == "left":
-        join_df = join_df.set_index("_key_left")
-        joined = (
-            left_df.merge(join_df, left_index=True, right_index=True, how="left")
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                how="left",
-                left_on="_key_right",
-                right_index=True,
-                suffixes=(None, None),
-            )
-            .drop(["_key_right"], axis=1)
-        )
+    original_length = len(right_df) if how == "right" else len(left_df)
+    (l_idx, r_idx), distances = _adjust_indexers(
+        indices, distances, original_length, how, predicate
+    )
+    # the `take` method doesn't allow introducing NaNs with -1 indices
+    # left = left_df.take(l_idx)
+    # therefore we are using the private _reindex_with_indexers as workaround
```

Comment:
How safe it is to rely on `_reindex_with_indexers`? Probably not very much given it's private, no?

Reply:
Yeah, safe for now I would say. But I was planning to open a pandas issue / PR to expose this publicly, and then once that is in we can update this code to use the public method for newer pandas